### PR TITLE
fixes #20694; range check the parameter of the proc `quit`

### LINKED ISSUE #20694: Invalid type in `importc: "exit"'
### What happened?

`exit` takes a `cint` (32-bit): https://man7.org/linux/man-pages/man3/exit.3.html

`quit` in Nim takes an `int` (64-bit): https://github.com/nim-lang/Nim/blob/534c97edc54ffb6d55baa5084f8c426a451c9244/lib/system.nim#L1136

in reality, only values 0-255 are well supported, but even if the full `cint` range was ok, a call to `quit(0x100000000)` will result in a `QuitSuccess` return code from the application after silent truncation.

### Nim Version

534c97edc54ffb6d55baa5084f8c426a451c9244

### Current Standard Output Logs

_No response_

### Expected Standard Output Logs

_No response_

### Possible Solution

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by ringabout**
fixes #20694

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -1087,32 +1087,7 @@ proc align(address, alignment: int): int =
     result = (address + (alignment - 1)) and not (alignment - 1)
 
 when defined(nimNoQuit):
-  proc quit*(errorcode: int = QuitSuccess) = discard "ignoring quit"
```
Comment (ringabout) [inline lib/system.nim:1090]: pre-existing documentation regression: https://github.com/nim-lang/Nim/commit/ce44cf03cc4a78741c423b2b3963b48b6d9e6755 I will fix it in the following PR.

File: lib/system.nim [hunk 2]
```diff
@@ -2289,6 +2260,57 @@ proc quit*(errormsg: string, errorcode = QuitFailure) {.noreturn.} =
...
+    systemEnv.rawQuit(errorcode)
+
+elif defined(js) and defined(nodejs) and not defined(nimscript):
+  proc quit*(errorcode: int = QuitSuccess) {.magic: "Exit",
+    importc: "process.exit", noreturn.}
+
+else:
+  proc quit*(errorcode: int = QuitSuccess) {.inline, noreturn.} =
+    when defined(posix): # posix uses low 8 bits
+      type exitCodeRange = int8
```
Comment (bung87) [inline lib/system.nim]: pascal case

File: lib/std/assertions.nim [hunk 3]
```diff
@@ -30,7 +28,7 @@ when not defined(nimHasSinkInference):
 
 proc raiseAssert*(msg: string) {.noinline, noreturn, nosinks.} =
   ## Raises an `AssertionDefect` with `msg`.
-  sysFatal(AssertionDefect, msg)
+  raise newException(AssertionDefect, msg)
```
Comment (arnetheduck) [inline lib/std/assertions.nim:31]: why is this change here? `sysFatal` afaik has magic meaning for the `--panics` dialect flag

File: lib/system.nim [hunk 4]
```diff
@@ -2285,16 +2246,66 @@ when defined(js):
...
+
+else:
+  proc quit*(errorcode: int = QuitSuccess) {.inline, noreturn.} =
+    when defined(posix): # posix uses low 8 bits
+      type ExitCodeRange = int8
+    else: # win32 uses low 32 bits
+      type ExitCodeRange = int32
+    let errorMsg = "forced to truncate exit code " & $errorcode & " to " & $(errorcode and high(ExitCodeRange))
+    if errorcode < low(ExitCodeRange) or errorcode > high(ExitCodeRange):
+      printErrorMessage errorMsg
```
Comment (arnetheduck) [inline lib/system.nim:2303]: quit shouldn't print to console, ever - this is not viable for many applications (that don't allocate a console at all, or where console output has strict formatting requirements) and doing so makes the entire quit mechanism broadly unusable

File: lib/std/assertions.nim (hunk 3 above)
Comment (ringabout) [reply lib/std/assertions.nim:31]: `std/assertions` has been moved out of system, but `sysFatal` is a private proc inside the system module. So I simply follow the action taken by `std/syncio` https://github.com/nim-lang/Nim/blob/59083e2e48bbf49daaa8d29e233c6f57f33a05ef/lib/std/syncio.nim#L154 

- including `system/fatal` doesn't work
- importing `system/fatal` doesn't work either (after making it importable, it causes a mysterious leaking with dollars)

File: lib/system.nim (hunk 4 above)
Comment (ringabout) [reply lib/system.nim:2303]: It has been pre-existing for quit with a string parameter.

Comment (arnetheduck) [reply lib/system.nim:2303]: it's not pre-existing for quit without a string parameter, which is what applications that don't write things to console use, for this exact reason: they don't want side effects of this sort

File: lib/std/assertions.nim (hunk 3 above)
Comment (ringabout) [reply lib/std/assertions.nim:31]: `raise newException` probably uses similar mechanism like `sysFatal` internally.

```nim
proc raiseExceptionAux(e: sink(ref Exception)) {.nodestroy.} =
  when defined(nimPanics):
    if e of Defect:
      reportUnhandledError(e)
      quit(1)
  ```

File: lib/system.nim (hunk 4 above)
Comment (arnetheduck) [reply lib/system.nim:2303]: the quit-with-string imo could be deprecated - the applications that want that can just `echo + quit` - it's of dubious utility in _system_ (cc @Araq )

File: lib/std/assertions.nim (hunk 3 above)
Comment (arnetheduck) [reply lib/std/assertions.nim:31]: the ability to raise panics that depend on the `panics` flag should be available for everyone I think - ie when I raise _any_ `Defect`, it should honor the panics flag - so the solution here would be that any time anything deriving from `Defect` is raised, it should use the equivalent of `sysFatal` - even if it's a "user-defined Defect"

changing `raiseAssert` significantly changes the way asserts behave, otherwise.

Comment (arnetheduck) [reply lib/std/assertions.nim:31]: > probably uses `sysFatal` internally

so maybe `sysFatal` needs to be removed entirely?

also, why is this a _runtime_ check? the type of the exception is usually known at `raise` time - there should not be runtime overhead for checking this.

Comment (ringabout) [reply lib/std/assertions.nim:31]: > changing raiseAssert significantly changes the way asserts behave, otherwise.

It still works like before

```nim
proc foo =
  raiseAssert("124")

proc main =
  try:
    foo()
  except AssertionDefect:
    discard

main()
```

with `panics`, `AssertionDefect` cannot be caught as before.

```
C:\Users\blue\Documents\GitHub\Nim\test2.nim(10) test2
C:\Users\blue\Documents\GitHub\Nim\test2.nim(6) main
C:\Users\blue\Documents\GitHub\Nim\test2.nim(2) foo
C:\Users\blue\Documents\GitHub\Nim\lib\std\assertions.nim(31) raiseAssert
Error: unhandled exception: 124 [AssertionDefect]
```

> also, why is this a runtime check? the type of the exception is usually known at raise time - there should not be runtime overhead for checking this.

That's a separate and pre-existing issue for both `sysFatal` and `raise`.

File: lib/system.nim (hunk 4 above)
Comment (Araq) [reply lib/system.nim:2303]: Why does `quit` have to be in system at all, string parameter or not.

Comment (c-blake) [reply lib/system.nim:2303]: Conceptually, `ansiC|os|osproc` do make sense, but I confess to not understanding the exact rules for slim system rearrangement..In any case, moving this one was not part of [The Roadmap](https://github.com/nim-lang/RFCs/issues/437). So, it has not seen broad exposure/discussion for feedback.  It does seem used 10+X less than `assert` in a quick survey of the nimbleverse.

Comment (arnetheduck) [reply lib/system.nim:2303]: > Why does quit have to be in system at all, string parameter or not.

`exitprocs` would be a candidate..

while we're here, it seems pretty reasonable to deprecate `programResult` - it's a pretty weird construct practically any part of the application can inject a return code that will not be seen until the program exits..

**Comment by Araq** [comment]
Very invasive change. Please implement my suggestion instead.

**Comment by ringabout** [comment]
Sure, I'm on it.

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/20753
